kmk/firmware.py

```diff
@@ -75,12 +83,83 @@ def _send_key(self, key):
             self._state.force_keycode_up(key)
             self._send_hid()
 
+    def _handle_update(self, update):
+        if update is not None:
+            self._state.matrix_changed(
+                update[0],
+                update[1],
+                update[2],
+            )
+
+            if self._state.hid_pending:
+                self._send_hid()
+
+            if self.debug_enabled:
+                print('New State: {}'.format(self._state._to_dict()))
+
+        self._state.process_timeouts()
+
+        for key in self._state.pending_keys:
+            self._send_key(key)
+            self._state.pending_key_handled()
+
+        if self._state.macro_pending:
+            for key in self._state.macro_pending(self):
+                self._send_key(key)
+
+            self._state.resolve_macro()
+
+    def _send_to_master(self, update):
+        if self.uart is not None:
+
+            if self.split_master_left:
+                update[1] += self.split_offsets[update[0]]
+
+            self.uart.write(update)
+
+    def _receive_from_slave(self):
+        if self.uart is not None and self.uart.in_waiting > 0:
+            update = bytearray(self.uart.read(3))
+            if self.split_master_left:
+                update[1] -= self.split_offsets[update[0]]
+            return update
+
+        return None
+
+    def _master_half(self):
+        return supervisor.runtime.serial_connected
+
+    def init_uart(self, tx=None, rx=None, timeout=1):
+        if self._master_half():
+            # If running with one wire, only receive on master
+            if rx is None or self.uart_flip:
+                return busio.UART(tx=rx, rx=None, timeout=timeout)
+            else:
+                return busio.UART(tx=tx, rx=rx, timeout=timeout)
+
+        else:
+            return busio.UART(tx=tx, rx=rx, timeout=timeout)
+
     def go(self):
         assert self.keymap, 'must define a keymap with at least one row'
         assert self.row_pins, 'no GPIO pins defined for matrix rows'
         assert self.col_pins, 'no GPIO pins defined for matrix columns'
         assert self.diode_orientation is not None, 'diode orientation must be defined'
 
+        if self.split_flip and not self._master_half():
+            self.col_pins = list(reversed(self.col_pins))
+
+        if self.split_side == "Left":
+            if supervisor.runtime.serial_connected:
```

Comment:
This can be flattened into `self.split_master_left = supervisor.runtime.serial_connected`, and the block of these below can be `self.split_master_left = not supervisor.runtime.serial_connected`

Reply:
It was left this way to leave room for bluetooth in the future.

Reply:
I retract this statement. It's being resolved.